dts: pm: Add wakeup-signal property

mmahadevan108 opened the pull request:
This is a new property to specify the wakeup signal number. There are some SoC's where the wakeup signal can be a different signal and not necessarily the interrupt number.
This property goes together with wakeup-source.

mmahadevan108 committed:
dts: pm: Add wakeup-signal property

This is a new property to specify the wakeup signal number.
There are some SoC's where the wakeup signal can be a different
signal and not necessarily the interrupt number.
This property goes together with wakeup-source.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

decsny reviewed (changes requested):
what would be the actual usage paradigm of this value?

mbolivar reviewed (changes requested):
I have the same question as @decsny . It feels like this is something that probably belongs in `interrupts`. The commit message doesn't make the example clear, but if the issue is that the "handler" for the wakeup signal is different than the interrupt controller for the "normal" interrupt in the IP blocks you're looking at, then I think this is probably a use case for `interrupts-extended` -- see the DT spec.

mmahadevan108 committed:
drivers: timer: Enable NXP LPTMR as wakeup-source

The NXP LPTMR uses the interrupt number or a dedicated
wakeup signal to wakeup the platform from low power
modes.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mbolivar commented:
@mmahadevan108 what hardware is responsible for handling the wakeup event? Is this a cortex M where it eventually has to go through the NVIC?

decsny commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -122,6 +122,10 @@ static int sys_clock_driver_init(void)
 
+#if (DT_INST_PROP(0, wakeup_source))
+	/* Use wakeup-signal property if specified, else use the irq number */
+	NXP_ENABLE_WAKEUP_SIGNAL(DT_INST_PROP_OR(0, wakeup_signal, DT_INST_IRQN(0)));
```
There is not enough value proposition being shown here to justify to be adding this property IMO. You are providing this value to a vendor specific macro. If we add this property it will become sticky and if people start using it and it turns out not to be very useful and becomes a technical debt when we decide some other way to do something, I don't want to add it. It's not thought through and seems rushed, and for what goal exactly I don't quite understand.

decsny replied:
Also I can see it being fulfilled by my https://github.com/zephyrproject-rtos/zephyr/pull/93838 if you think of it as a one bit multiplexor

mbolivar commented:
@DerekSnell I appreciate the explanation, thanks.

From what you have said, my instincts would be to model the WUU as a devicetree interrupt controller, and treat peripherals that are capable of asynchronously signaling the WUU on an event as interrupt-generating devices whose interrupts are handled by the WUU node. If they can also generate interrupts that go to the NVIC, that is a use case for `interrupts-extended`. The fact that the WUU can in turn interrupt the NVIC is already supported by zephyr's multi-level interrupt infrastructure as far as I can see.

Please see section 2.4 in the devicetree specification if you are unfamiliar with the details on how interrupts work in DT.

Of course, see https://docs.zephyrproject.org/latest/kernel/services/interrupts.html for details on the interrupt controller hierarchy in zephyr.

See https://github.com/zephyrproject-rtos/zephyr/tree/main/dts/riscv/openisa for some in-tree examples of SoCs that stitch all this together in DT, and you can grep around in the sources to see example drivers that use these multi level IRQ controller nodes in practice.

If there is really something missing here in the existing interrupt handling infrastructure that you can articulate after familiarizing yourself with the DT and Zephyr standard way of doing this, please let me know and we can continue discussion. Absent something convincing, though, this PR feels like it's reinventing the wheel.

mbolivar commented:
> But as far as I can tell there is no standardized interrupt interface of zephyr, all the interrupt controller drivers are vendor specific interfaces

What do you mean? See drivers/interrupt_controller for the implementation of the standard interfaces. This is how Zephyr makes IRQ_CONNECT work with multi level interrupt handlers.
